packages/microcosm-dom/src/presenter.js

```diff
@@ -0,0 +1,172 @@
+import { Microcosm, Observable, tag } from 'microcosm'
+import { advice, noop } from './utilities'
+
+export function generatePresenter(createElement, Component) {
+  function renderMediator() {
+    return createElement(PresenterMediator, {
+      repo: this.props.repo,
+      presenter: this,
+      __props: this.props,
+      __state: this.state
+    })
+  }
+
+  class Presenter extends Component {
+    constructor() {
+      super(...arguments)
+
+      this.send = this.send.bind(this)
+      this.state = {}
+
+      // Ensure key lifecycle methods are protected by first applying
+      // prototype behavior, thenany extended behavior
+      advice(Presenter, this, 'componentWillUpdate')
+      advice(Presenter, this, 'componentDidMount')
+
+      // Override the given render method so that the mediator can boot.
+      // The original render method gets called in the mediator
+      this.render = renderMediator
+    }
+
+    get model() {
+      return this.mediator.model.payload || {}
+    }
+
+    get repo() {
+      return this.mediator.repo
+    }
+
+    setup(repo, props, state) {
+      // NOOP
+    }
+
+    ready(repo, props, state) {
+      // NOOP
+    }
+
+    update(repo, nextProps, nextState) {
+      // NOOP
+    }
+
+    teardown(repo, props, state) {
+      // NOOP
+    }
+
+    intercept() {
+      return {}
+    }
+
+    getRepo(inherited) {
+      return inherited ? inherited.fork() : new Microcosm()
+    }
+
+    getModel(repo, presenterProps, presenterState) {
+      return {}
+    }
+
+    componentDidMount() {
+      this.ready(this.repo, this.props, this.state)
+    }
+
+    componentWillUpdate(props, state) {
+      this.mediator.updateModel(props, state)
+      this.update(this.repo, props, state)
+    }
+
+    componentWillUnmount() {
+      this.mediator.model.cancel()
+      this.teardown(this.repo, this.props, this.state)
+
+      if (this.didFork) {
+        this.repo.complete()
+      }
+    }
+
+    render() {
+      if (Array.isArray(this.props.children)) {
+        return this.props.children[0] || null
```

Comment:
This is the only really nasty wart. Preact sets children to an array _always_, where as React does not. Both must return null from render (I think this is true of React, but definitely Preact).

React can also return an array from render as of 16.